### PR TITLE
Batching rules for complex view functions

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47227 Batched gradient support for view+inplace operations
* #47365 Add batching rule for torch.clone(tensor, torch.contiguous_format)
* #47364 as_strided batching rule
* #47226 Batching rule for Tensor.new_empty_strided
* #47225 Implement Tensor.new_empty_strided(sizes, strides, *, dtype, device, requires_grad)
* #47223 Batched grad for advanced indexing (index)
* #47189 Fix sum batching rule, add simple clone batching rule
* **#47188 Batching rules for complex view functions**

Includes batching rules for:
- torch.real, torch.imag, torch.view_as_real, and torch.view_as_complex

Test Plan:
- new tests

Differential Revision: [D24741686](https://our.internmc.facebook.com/intern/diff/D24741686)

### CODE REVIEW

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 1]
```diff
@@ -337,13 +337,18 @@ Tensor view_batching_rule(const Tensor& self, IntArrayRef size) {
...
 template <typename F, F Func, typename... ExtraArgs>
-Tensor unary_pointwise_batching_rule(const Tensor& input, ExtraArgs... args) {
+Tensor unwrap_and_call(const Tensor& input, ExtraArgs... args) {
   auto* input_batched = unsafeGetBatchedImpl(input);
   auto output_physical = Func(input_batched->value(), args...);
   auto old_bdims = input_batched->bdims();
   return makeBatched(output_physical, BatchDims(old_bdims.begin(), old_bdims.end()));
 }
 
+template <typename F, F Func, typename... ExtraArgs>
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:347]: I am not sure to understand why you need this extra function? How is it different from the one above?

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 2]
```diff
@@ -683,6 +687,17 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
 #undef BINARY_POINTWISE_VA
 #undef BINARY_POINTWISE
 
+
+#define TRIVIAL_OP(op) m.impl(#op, \
+    unwrap_and_call<Tensor (*)(const Tensor&), at::op>);
+  // complex number view operators
+  TRIVIAL_OP(imag)
+  TRIVIAL_OP(conj);
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:695]: conj is not a view for complex Tensor?

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 3]
```diff
@@ -683,6 +687,17 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
...
+  TRIVIAL_OP(real);
+  TRIVIAL_OP(view_as_complex);
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:697]: Will this one work fine if the batch dimension is the last one?
This case is explicitly avoided in the tests with the `shape_postfix` argument.

File: test/test_vmap.py [hunk 4]
```diff
@@ -1210,6 +1210,45 @@ def test_expand_as(self):
...
+    def _test_complex_views(self, op, dtypes, shape_postfix=()):
+        test = self._vmap_view_test
+
+        def run_test(op, dtype):
+            def get(shape):
+                shape = shape + list(shape_postfix)
+                return torch.randn(shape, dtype=dtype)
+
+            B0, B1 = 7, 11
+            test = self._vmap_view_test
```
Comment (albanD) [inline test/test_vmap.py:1222]: nit: this is already defined above

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:347]: The only thing that is different is th ename. I thought it would be strange if we used "unary_pointwise_batching_rule" to implement batching rules for operations that aren't unary pointwise (e.g., the complex view functions aren't unary pointwise).

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:695]: Thanks for the catch! I went through this too quickly

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 3 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:697]: That's a really good point. No, this won't work fine in that case. We'll need a different batching rule here (to move the batch dim to the front).

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:347]: But I guess I could just name "unary_pointwise_batching_rule" to "unwrap_and_call"

Comment (albanD) [reply aten/src/ATen/BatchingRegistrations.cpp:347]: Ho ok. I was looking very hard for some subtle template differences.
I guess that using `unwrap_and_call` is better.

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 3 above)
Comment (albanD) [reply aten/src/ATen/BatchingRegistrations.cpp:697]: But will this movement of dimension be an issue for view operations? As the move might incur a copy?
Should this just be forbidden? (the last dimension has to be the 2 complex values and cannot be a batch dim)

Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:697]: > But will this movement of dimension be an issue for view operations? As the move might incur a copy?

The move won't incur a copy (movement is just a permute).

> Should this just be forbidden? (the last dimension has to be the 2 complex values and cannot be a batch dim)

`view_as_complex` requires that the last dimension has [stride 1 for the view to be valid](https://github.com/pytorch/pytorch/blob/a341a4329a0dd625d890212d72bbd62d3840ba0a/aten/src/ATen/native/ComplexHelper.h#L66) (because we're just viewing [[a0 b0], [a1 b1]] as [a0 b0 a1 b1]. So if the batch dim is at the end of the tensor, we can support `view_as_complex` depending on the strides of the dimension of size 2

```
# This should fail gracefully because the dimension of size 2 doesn't have stride 1.
x = torch.randn(2, B0)
vmap(torch.view_as_complex, in_dim=1)(x)

# This will succeed 
x = torch.randn(B0, 2).t()
vmap(torch.view_as_complex, in_dim=1)(x)

**Review by albanD — APPROVED** [review]
lgtm
